pkg/cloud/mesos/mesos.go

```diff
@@ -14,67 +15,118 @@ import (
 )
 
 var (
-	PluginName = "mesos"
+	PluginName    = "mesos"
+	CloudProvider *MesosCloud
 
+	initialized         = make(chan struct{})
 	noHostNameSpecified = errors.New("No hostname specified")
-
-	//TODO(jdef) this should handle mesos upid's (perhaps once we move to pure bindings)
-	mesosMaster = flag.String("mesos_master", "localhost:5050", "Location of leading Mesos master. Default localhost:5050.")
 )
 
 func init() {
 	cloudprovider.RegisterCloudProvider(
 		PluginName,
-		func(conf io.Reader) (cloudprovider.Interface, error) {
-			return newMesosCloud()
+		func(configReader io.Reader) (cloudprovider.Interface, error) {
+			provider, err := newMesosCloud(configReader)
+			if CloudProvider != nil {
+				return CloudProvider, nil
+			}
+			if err == nil {
+				CloudProvider = provider
+				close(initialized)
+			}
+			return provider, err
 		})
 }
 
 type MesosCloud struct {
 	client *mesosClient
+	config *Config
+}
+
+// Blocks until the Mesos cloud provider has been initialized via the factory
+// function defined at registration time.
+func AwaitInitialization() {
```

Comment:
maybe this would be more aptly named CheckInitialized():
```go
func CheckInitialized() {
  select {
    case <-initialized:
      return
    default:
      //programmer error
      panic("missing prior call to cloudprovider.InitCloudProvider for 'mesos' cloud provider")
  }
}